desktop/core/src/desktop/lib/fs/gc/gs.py

```diff
@@ -0,0 +1,454 @@
+        raise e
+
+    if key:
+      key.set_contents_from_string(data or '', replace=overwrite)
+    else:
+      raise Exception('Cannot perform create operation.')
+
+  def _get_key(self, path, validate=True):
+    bucket_name, key_name = parse_uri(path)[:2]
+    bucket = self._get_bucket(bucket_name)
+
+    try:
+      return bucket.get_key(key_name, headers=self.header_values)
+    except BotoClientError as e:
+      raise GSFileSystemException(_('Failed to access path at "%s": %s') % (path, e.reason))
+    except GSResponseError as e:
+      if e.status in (301, 400):
+        raise GSFileSystemException(_('Failed to access path: "%s" '
+          'Check that you have access to read this bucket and that the region is correct: %s') % (path, e.message or e.reason))
+      elif e.status == 403:
+        raise GSFileSystemException(_('User is not authorized to access path at "%s".' % path))
+      else:
+        raise GSFileSystemException(e.message or e.reason)
+    except GSResponseError as e:
+      raise e
+
+  @translate_gs_error
+  def open(self, path, mode='r'):
+    key = self._get_key(path)
+    if key is None:
+      raise GSFileSystemException("No such file or directory: '%s'" % path)
+    return gsfile_open(key, mode=mode)
+
+  @translate_gs_error
+  def listdir_stats(self, path, glob=None):
+    """List and get stats for files and directories in a GS bucket.
+    For path 'gs://', it gets stats for all listed buckets in GS filesystem.
+
+    Args:
+      path (str): The GS path to list.
+      glob (str, optional): Glob pattern for filtering files. Default is None.
+
+    Returns:
+      list of GSStat: A list of GSStat objects representing files and directories in the path.
+                      For 'gs://' path, it return a list of GSStat objects for all listed buckets.
+    """
+    if glob is not None:
+      raise NotImplementedError(_("Option `glob` is not implemented"))
+
+    if GSFileSystem.isroot(path):
+      # Return sorted stats of all listed buckets for path gs://
+      try:
+        return sorted(
+          [GSStat.from_bucket(b, self.fs) for b in self._s3_connection.get_all_buckets(headers=self.header_values)], key=lambda x: x.name)
+      except GSFileSystemException as e:
+        raise e
+      except GSResponseError as e:
+        if 'Forbidden' in str(e) or (hasattr(e, 'status') and e.status == 403):
+          raise GSListAllBucketsException(
+            _('You do not have permissions to list all buckets. Please specify a bucket name you have access to.'))
+        else:
+          raise GSFileSystemException(_('Failed to retrieve buckets: %s') % e.reason)
+      except Exception as e:
+        raise GSFileSystemException(('Failed to retrieve buckets: %s') % e)
+
+    bucket_name, prefix = parse_uri(path)[:2]
+    bucket = self._get_bucket(bucket_name)
+    prefix = self._append_separator(prefix)
+
+    res = []
+    for item in bucket.list(prefix=prefix, delimiter='/', headers=self.header_values):
```

Comment:
Do we normally send the full list of files here in a directory without any pagination? Getting the stats for each of the might become a bottleneck if there are thousands of these.

There is no need to make the change here in this PR if we are doing the same for s3.

Reply:
Yes we are getting the full list of keys with keyname prefix and its the same for S3. I guess this method's role is to just fetch the stats for each FS and then pagination is done by Hue here: https://github.com/cloudera/hue/blob/cd0f6bf7b62114e33f9424121ec4c2b776b51ff8/apps/filebrowser/src/filebrowser/views.py#L509

Reply:
What can be the bottleneck here @amitsrivastava? The loop running on the fetched key list?

Reply:
yes, making individual calls for potentially thousands of files to get their stats could be problematic.

Reply:
IIUC we are making only 1 call to fetch the list of keys and then creating the stats objects using the for loop.